ProjFS.Mac/PrjFSKext/VnodeCache.cpp

```diff
@@ -0,0 +1,414 @@
+KEXT_STATIC uintptr_t s_ModBitmask;
+
+static RWLock s_entriesLock;
+
+kern_return_t VnodeCache_Init()
+{
+    if (RWLock_IsValid(s_entriesLock))
+    {
+        return KERN_FAILURE;
+    }
+    
+    s_entriesLock = RWLock_Alloc();
+    if (!RWLock_IsValid(s_entriesLock))
+    {
+        return KERN_FAILURE;
+    }
+
+    s_entriesCapacity = ComputePow2CacheCapacity(desiredvnodes);
+    s_ModBitmask = s_entriesCapacity - 1;
+    
+    s_entries = Memory_AllocArray<VnodeCacheEntry>(s_entriesCapacity);
+    if (nullptr == s_entries)
+    {
+        s_entriesCapacity = 0;
+        return KERN_RESOURCE_SHORTAGE;
+    }
+    
+    memset(s_entries, 0, s_entriesCapacity * sizeof(VnodeCacheEntry));
+    
+    PerfTracing_RecordSample(PrjFSPerfCounter_CacheCapacity, 0, s_entriesCapacity);
+    
+    return KERN_SUCCESS;
+}
+
+kern_return_t VnodeCache_Cleanup()
+{
+    if (nullptr != s_entries)
+    {
+        Memory_FreeArray<VnodeCacheEntry>(s_entries, s_entriesCapacity);
+        s_entries = nullptr;
+        s_entriesCapacity = 0;
+    }
+    
+    if (RWLock_IsValid(s_entriesLock))
+    {
+        RWLock_FreeMemory(&s_entriesLock);
+        return KERN_SUCCESS;
+    }
+    
+    return KERN_FAILURE;
+}
+
+VirtualizationRootHandle VnodeCache_FindRootForVnode(
+    PerfTracer* _Nonnull perfTracer,
+    PrjFSPerfCounter cacheHitCounter,
+    PrjFSPerfCounter cacheMissCounter,
+    PrjFSPerfCounter cacheMissFallbackFunctionCounter,
+    PrjFSPerfCounter cacheMissFallbackFunctionInnerLoopCounter,
+    vnode_t _Nonnull vnode,
+    vfs_context_t _Nonnull context)
+{
+    VirtualizationRootHandle rootHandle = RootHandle_None;
+    uintptr_t vnodeHashIndex = ComputeVnodeHashIndex(vnode);
+    uint32_t vnodeVid = vnode_vid(vnode);
+    
+    if (TryGetVnodeRootFromCache(vnode, vnodeHashIndex, vnodeVid, rootHandle))
+    {
+        perfTracer->IncrementCount(cacheHitCounter, true /*ignoreSampling*/);
+        return rootHandle;
+    }
+    
+    perfTracer->IncrementCount(cacheMissCounter, true /*ignoreSampling*/);
+    FindVnodeRootFromDiskAndUpdateCache(
+        perfTracer,
+        cacheMissFallbackFunctionCounter,
+        cacheMissFallbackFunctionInnerLoopCounter,
+        context,
+        vnode,
+        vnodeHashIndex,
+        vnodeVid,
+        UpdateCacheBehavior_TrustCurrentEntry,
+        rootHandle);
+    
+    return rootHandle;
+}
+
+VirtualizationRootHandle VnodeCache_RefreshRootForVnode(
+    PerfTracer* _Nonnull perfTracer,
+    PrjFSPerfCounter cacheHitCounter,
+    PrjFSPerfCounter cacheMissCounter,
+    PrjFSPerfCounter cacheMissFallbackFunctionCounter,
+    PrjFSPerfCounter cacheMissFallbackFunctionInnerLoopCounter,
+    vnode_t _Nonnull vnode,
+    vfs_context_t _Nonnull context)
+{
+    VirtualizationRootHandle rootHandle = RootHandle_None;
+    uintptr_t vnodeHashIndex = ComputeVnodeHashIndex(vnode);
+    uint32_t vnodeVid = vnode_vid(vnode);
+    
+    perfTracer->IncrementCount(cacheMissCounter, true /*ignoreSampling*/);
```

Comment:
Are we purposely making this a cache miss, because we're refreshing the value?
If so i'm wondering if we want to capture it another way as well in the future

Reply:
> Are we purposely making this a cache miss, because we're refreshing the value?

Yeah that's the idea.  Currently "cache miss" is recorded anytime the cache does not have the root we're looking for.

> If so i'm wondering if we want to capture it another way as well in the future

Agreed, I will keep this in mind while working on #973